##  DAY 10
**Data wrangling**

What is data wrangling?

Data wrangling is the process of converting raw data into a usable form.\
It may also be called data mugging or data remediation.

Data wrangling describes a series of processes designed to explore, transform, and validate raw datasets from their messy and complex forms into high-quality data.\
You can use your wrangled data to produce valuable insights and guide business decisions. 

**DATA  WRANGLING STEPS**

-Discovery\
-Transformation\
-Validation\
-Publishing

What are the steps in data wrangling which is learned by me?
> for me\
> check the data what it is about \
> getter information about the dataset even from the internet\
> see the columns , datatypes\
> then see if the columns is well named\
> see the null values\
> check the outliers \
> remove the outliers\
> replace the null values\
> if there is something wrong in the scale of calculation make sure you make it right\
> you data shall be cleaned after that \
> check your data one last time\
> publish

In [86]:
import pandas as pd
import numpy as np
import seaborn as sns

In [87]:
df = pd.read_csv("summer_movies.csv")

In [88]:
df.head()

tconst title_type                    primary_title  \
0  tt0011462      movie                Midsummer Madness   
1  tt0026714      movie        A Midsummer Night's Dream   
2  tt0033864      movie  The Teachers on Summer Vacation   
3  tt0037325      movie                     Summer Storm   
4  tt0038406      movie                Centennial Summer   

              original_title    year  runtime_minutes                  genres  \
0          Midsummer Madness  1920.0             60.0                   Drama   
1  A Midsummer Night's Dream  1935.0            133.0  Comedy,Fantasy,Romance   
2   Magistrarna på sommarlov  1941.0             86.0                  Comedy   
3               Summer Storm  1944.0            106.0   Crime,Drama,Film-Noir   
4          Centennial Summer  1946.0            102.0   History,Music,Romance   

                      simple_title  average_rating  num_votes  
0                midsummer madness             7.4         19  
1         a midsummer nights dream             6.8       3931  
2  the teachers on summer vacation             5.5         78  
3                     summer storm             6.6        688  
4                centennial summer             6.1        431

removing the useless columns

In [89]:
df.drop(columns='primary_title', inplace=True)
df.drop(columns='simple_title', inplace=True)



In [90]:
df.drop(columns='tconst', inplace=True)

In [91]:
df

title_type                      original_title    year  runtime_minutes  \
0        movie                   Midsummer Madness  1920.0             60.0   
1        movie           A Midsummer Night's Dream  1935.0            133.0   
2        movie            Magistrarna på sommarlov  1941.0             86.0   
3        movie                        Summer Storm  1944.0            106.0   
4        movie                   Centennial Summer  1946.0            102.0   
..         ...                                 ...     ...              ...   
900      movie                         Summer Love  2019.0              NaN   
901      movie   Lass den Sommer nie wieder kommen  2017.0            202.0   
902      movie                      Goodbye Summer  2019.0             67.0   
903      movie                      Summer of Mesa  2020.0             76.0   
904      movie  Ein wilder Sommer - Die Wachausaga  2018.0            154.0   

                     genres  average_rating  num_votes  
0                     Drama             7.4         19  
1    Comedy,Fantasy,Romance             6.8       3931  
2                    Comedy             5.5         78  
3     Crime,Drama,Film-Noir             6.6        688  
4     History,Music,Romance             6.1        431  
..                      ...             ...        ...  
900                 Romance             7.5         17  
901                   Drama             6.2         98  
902           Drama,Romance             5.6         79  
903           Drama,Romance             5.5        114  
904    Comedy,Drama,History             4.4         15  

[905 rows x 7 columns]

**Basic data exploration**

In [92]:
df.head(1)

title_type     original_title    year  runtime_minutes genres  \
0      movie  Midsummer Madness  1920.0             60.0  Drama   

   average_rating  num_votes  
0             7.4         19

In [93]:
df.tail(3)

title_type                      original_title    year  runtime_minutes  \
902      movie                      Goodbye Summer  2019.0             67.0   
903      movie                      Summer of Mesa  2020.0             76.0   
904      movie  Ein wilder Sommer - Die Wachausaga  2018.0            154.0   

                   genres  average_rating  num_votes  
902         Drama,Romance             5.6         79  
903         Drama,Romance             5.5        114  
904  Comedy,Drama,History             4.4         15

In [94]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 905 entries, 0 to 904
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title_type       905 non-null    object 
 1   original_title   905 non-null    object 
 2   year             904 non-null    float64
 3   runtime_minutes  849 non-null    float64
 4   genres           887 non-null    object 
 5   average_rating   905 non-null    float64
 6   num_votes        905 non-null    int64  
dtypes: float64(3), int64(1), object(3)
memory usage: 49.6+ KB


In [95]:
df.describe()

year  runtime_minutes  average_rating      num_votes
count   904.000000       849.000000      905.000000     905.000000
mean   2001.353982        91.043581        6.273702    2296.285083
std      19.448166        23.854447        1.245414   20630.393488
min    1920.000000         5.000000        2.300000      10.000000
25%    1990.000000        82.000000        5.600000      34.000000
50%    2008.000000        91.000000        6.400000     119.000000
75%    2016.000000       101.000000        7.100000     517.000000
max    2024.000000       237.000000        9.800000  564894.000000

In [96]:
df.shape

(905, 7)

In [97]:
df.rename(columns={
    'runtime_minutes': 'Runtime',
    'average_rating' : 'Rating',
    'num_votes' : 'Votes',
    'original_title' : 'Title'
},inplace=True)

In [98]:
df

title_type                               Title    year  Runtime  \
0        movie                   Midsummer Madness  1920.0     60.0   
1        movie           A Midsummer Night's Dream  1935.0    133.0   
2        movie            Magistrarna på sommarlov  1941.0     86.0   
3        movie                        Summer Storm  1944.0    106.0   
4        movie                   Centennial Summer  1946.0    102.0   
..         ...                                 ...     ...      ...   
900      movie                         Summer Love  2019.0      NaN   
901      movie   Lass den Sommer nie wieder kommen  2017.0    202.0   
902      movie                      Goodbye Summer  2019.0     67.0   
903      movie                      Summer of Mesa  2020.0     76.0   
904      movie  Ein wilder Sommer - Die Wachausaga  2018.0    154.0   

                     genres  Rating  Votes  
0                     Drama     7.4     19  
1    Comedy,Fantasy,Romance     6.8   3931  
2                    Comedy     5.5     78  
3     Crime,Drama,Film-Noir     6.6    688  
4     History,Music,Romance     6.1    431  
..                      ...     ...    ...  
900                 Romance     7.5     17  
901                   Drama     6.2     98  
902           Drama,Romance     5.6     79  
903           Drama,Romance     5.5    114  
904    Comedy,Drama,History     4.4     15  

[905 rows x 7 columns]

>The data looks good now

#### data selection and cleaning

Now finding the missing value

In [99]:
df.isnull().sum()

title_type     0
Title          0
year           1
Runtime       56
genres        18
Rating         0
Votes          0
dtype: int64

In [100]:
#finding the average of null values
df.isnull().sum() * 100 / len(df)

title_type    0.000000
Title         0.000000
year          0.110497
Runtime       6.187845
genres        1.988950
Rating        0.000000
Votes         0.000000
dtype: float64

In [101]:
df.iloc[888:895]

title_type                                              Title    year  \
888      video  'Hocus Pocus' 25th Anniversary Panel at Midsummer  2018.0   
889      movie                                        Summer Camp  2018.0   
890      movie                                        Zulu Summer  2019.0   
891      movie                               This Guest of Summer     NaN   
892      movie                                The Girls of Summer  2020.0   
893      movie                                  60 Days of Summer  2018.0   
894      movie                                     Passing Summer  2018.0   

     Runtime        genres  Rating  Votes  
888     80.0     Talk-Show     7.8     16  
889    128.0  Comedy,Drama     7.9     16  
890     78.0   Documentary     8.0     52  
891      NaN        Horror     7.9    179  
892     94.0         Drama     5.9    101  
893     94.0         Drama     6.2     40  
894     93.0       Romance     6.0     40

knowing the location of the null values in the  column

In [102]:
df.sort_values(by='year' ,na_position='first')

title_type                      Title    year  Runtime  \
891      movie       This Guest of Summer     NaN      NaN   
0        movie          Midsummer Madness  1920.0     60.0   
207      movie            Her Summer Hero  1928.0     60.0   
1        movie  A Midsummer Night's Dream  1935.0    133.0   
2        movie   Magistrarna på sommarlov  1941.0     86.0   
..         ...                        ...     ...      ...   
700      movie          Griffin in Summer  2024.0     90.0   
641      movie           Mezhdu nami leto  2024.0    113.0   
704      movie        Summers, el rebelde  2024.0     83.0   
721      movie        Tabestane Haman Sal  2024.0      NaN   
719      movie     Quell'estate con Irène  2024.0     90.0   

                     genres  Rating  Votes  
891                  Horror     7.9    179  
0                     Drama     7.4     19  
207                  Comedy     4.3     12  
1    Comedy,Fantasy,Romance     6.8   3931  
2                    Comedy     5.5     78  
..                      ...     ...    ...  
700                  Comedy     8.6     38  
641    Comedy,Drama,Musical     5.9     42  
704             Documentary     7.7     14  
721                     NaN     8.2    116  
719                   Drama     6.2     39  

[905 rows x 7 columns]

dropping the null value of a single column

In [103]:
df = df.dropna(subset='year')

In [114]:
df.isnull().sum()

title_type    0
Title         0
year          0
Rating        0
Votes         0
dtype: int64

dropping the null values of all columns

In [115]:
df = df.dropna(axis=1)

count

In [106]:
df.count()

title_type    904
Title         904
year          904
Rating        904
Votes         904
dtype: int64

after the removing the null values checking again

In [107]:
df.isnull().sum()

title_type    0
Title         0
year          0
Rating        0
Votes         0
dtype: int64

checking the null value of index 891 if its is removed or not

In [108]:
df.iloc[890]

title_type          movie
Title         Zulu Summer
year               2019.0
Rating                8.0
Votes                  52
Name: 890, dtype: object

In [109]:
df.nunique()

title_type      3
Title         837
year           83
Rating         70
Votes         487
dtype: int64

In [110]:
df.head()

title_type                      Title    year  Rating  Votes
0      movie          Midsummer Madness  1920.0     7.4     19
1      movie  A Midsummer Night's Dream  1935.0     6.8   3931
2      movie   Magistrarna på sommarlov  1941.0     5.5     78
3      movie               Summer Storm  1944.0     6.6    688
4      movie          Centennial Summer  1946.0     6.1    431

In [111]:
pd.pivot_table(df, values='year', index='Title', 
columns='Rating')

Rating                                              2.3  2.4  2.5  2.7  2.9  \
Title                                                                         
#MST3KMindlessSummer Livestream III: BLOOD WATE...  NaN  NaN  NaN  NaN  NaN   
'Hocus Pocus' 25th Anniversary Panel at Midsummer   NaN  NaN  NaN  NaN  NaN   
(500) Days of Summer                                NaN  NaN  NaN  NaN  NaN   
...som havets nakna vind                            NaN  NaN  NaN  NaN  NaN   
10:30 P.M. Summer                                   NaN  NaN  NaN  NaN  NaN   
...                                                 ...  ...  ...  ...  ...   
Zulu Summer                                         NaN  NaN  NaN  NaN  NaN   
Zwischen Sommer und Herbst                          NaN  NaN  NaN  NaN  NaN   
a-ha: MTV Unplugged - Summer Solstice               NaN  NaN  NaN  NaN  NaN   
Äntligen midsommar!                                 NaN  NaN  NaN  NaN  NaN   
Été 85                                              NaN  NaN  NaN  NaN  NaN   

Rating                                              3.0  3.1  3.2  3.3  3.4  \
Title                                                                         
#MST3KMindlessSummer Livestream III: BLOOD WATE...  NaN  NaN  NaN  NaN  NaN   
'Hocus Pocus' 25th Anniversary Panel at Midsummer   NaN  NaN  NaN  NaN  NaN   
(500) Days of Summer                                NaN  NaN  NaN  NaN  NaN   
...som havets nakna vind                            NaN  NaN  NaN  NaN  NaN   
10:30 P.M. Summer                                   NaN  NaN  NaN  NaN  NaN   
...                                                 ...  ...  ...  ...  ...   
Zulu Summer                                         NaN  NaN  NaN  NaN  NaN   
Zwischen Sommer und Herbst                          NaN  NaN  NaN  NaN  NaN   
a-ha: MTV Unplugged - Summer Solstice               NaN  NaN  NaN  NaN  NaN   
Äntligen midsommar!                                 NaN  NaN  NaN  NaN  NaN   
Été 85                                              NaN  NaN  NaN  NaN  NaN   

Rating                                              ...  8.5     8.6  8.7  \
Title                                               ...                     
#MST3KMindlessSummer Livestream III: BLOOD WATE...  ...  NaN     NaN  NaN   
'Hocus Pocus' 25th Anniversary Panel at Midsummer   ...  NaN     NaN  NaN   
(500) Days of Summer                                ...  NaN     NaN  NaN   
...som havets nakna vind                            ...  NaN     NaN  NaN   
10:30 P.M. Summer                                   ...  NaN     NaN  NaN   
...                                                 ...  ...     ...  ...   
Zulu Summer                                         ...  NaN     NaN  NaN   
Zwischen Sommer und Herbst                          ...  NaN     NaN  NaN   
a-ha: MTV Unplugged - Summer Solstice               ...  NaN  2017.0  NaN   
Äntligen midsommar!                                 ...  NaN     NaN  NaN   
Été 85                                              ...  NaN     NaN  NaN   

Rating                                              8.8  8.9  9.0  9.1  9.2  \
Title                                                                         
#MST3KMindlessSummer Livestream III: BLOOD WATE...  NaN  NaN  NaN  NaN  NaN   
'Hocus Pocus' 25th Anniversary Panel at Midsummer   NaN  NaN  NaN  NaN  NaN   
(500) Days of Summer                                NaN  NaN  NaN  NaN  NaN   
...som havets nakna vind                            NaN  NaN  NaN  NaN  NaN   
10:30 P.M. Summer                                   NaN  NaN  NaN  NaN  NaN   
...                                                 ...  ...  ...  ...  ...   
Zulu Summer                                         NaN  NaN  NaN  NaN  NaN   
Zwischen Sommer und Herbst                          NaN  NaN  NaN  NaN  NaN   
a-ha: MTV Unplugged - Summer Solstice               NaN  NaN  NaN  NaN  NaN   
Äntligen midsommar!                                 NaN  NaN  NaN  NaN  N

In [112]:
df.head(2)

title_type                      Title    year  Rating  Votes
0      movie          Midsummer Madness  1920.0     7.4     19
1      movie  A Midsummer Night's Dream  1935.0     6.8   3931

In [113]:
df.duplicated().sum()

np.int64(0)